File: conductor/rate_limit.py
"""Rate-limit token bucket — Lua-backed, single-key, cluster-safe.

The Lua script is loaded once per Redis connection-pool via SCRIPT LOAD
(redis-py's `register_script`); subsequent calls are EVALSHA. The wrapper
keeps the call site simple: pass site, queue, and bucket parameters; get
back `(allowed, retry_after_ms)`.

Time is provided as `now_ms` explicitly (rather than reading from Redis's
TIME) so unit tests can drive the clock deterministically. Production
callers pass `int(time.time() * 1000)`.
"""

from __future__ import annotations

import threading
from pathlib import Path

import redis as redis_mod

_SCRIPT_PATH = Path(__file__).with_name("rate_limit.lua")
_LUA_SOURCE: str | None = None
_LUA_LOCK = threading.Lock()
# Per-Redis-client `Script` cache — re-using the same registered Script object
# across calls means redis-py reuses its SHA1 and only EVALSHAs.
_REGISTERED: dict[int, "redis_mod.client.Script"] = {}


def rate_key(site: str, queue: str) -> str:
    return f"conductor:{site}:rate:{queue}"


def _get_script(client: redis_mod.Redis) -> "redis_mod.client.Script":
    global _LUA_SOURCE
    if _LUA_SOURCE is None:
        with _LUA_LOCK:
            if _LUA_SOURCE is None:
                _LUA_SOURCE = _SCRIPT_PATH.read_text(encoding="utf-8")
    cid = id(client.connection_pool)
    script = _REGISTERED.get(cid)
    if script is None:
        script = client.register_script(_LUA_SOURCE)
        _REGISTERED[cid] = script
    return script


def take_token(
    client: redis_mod.Redis,
    site: str,
    queue: str,
    *,
    max_tokens: int,
    refill_per_sec: float,
    now_ms: int,
    n: int = 1,
) -> tuple[bool, int]:
    """Try to consume `n` tokens. Returns (allowed, retry_after_ms).
    If allowed is False, retry_after_ms is the minimum wait before this
    many tokens will be available."""
    script = _get_script(client)
    out = script(
        keys=[rate_key(site, queue)],
        args=[max_tokens, refill_per_sec, now_ms, n],
    )
    # redis-py returns Lua tables as Python lists.
    allowed = bool(int(out[0]))
    retry_ms = int(out[1])
    return allowed, retry_ms

File: conductor/rate_limit.lua
-- take_token: atomic refill bucket.
-- KEYS[1] = "conductor:{site}:rate:{queue}"
-- ARGV    = max_tokens, refill_per_sec, now_ms, n
-- returns: {allowed (0|1), retry_after_ms}

local max_tokens     = tonumber(ARGV[1])
local refill_per_sec = tonumber(ARGV[2])
local now_ms         = tonumber(ARGV[3])
local n              = tonumber(ARGV[4])

local state    = redis.call("HMGET", KEYS[1], "tokens", "last_refill_ms")
local tokens   = tonumber(state[1])
local last_ms  = tonumber(state[2])

-- First call (key missing): start full.
if tokens == nil then
    tokens  = max_tokens
    last_ms = now_ms
end

local elapsed_ms = now_ms - last_ms
if elapsed_ms < 0 then elapsed_ms = 0 end

local refill = (elapsed_ms * refill_per_sec) / 1000.0
tokens = tokens + refill
if tokens > max_tokens then tokens = max_tokens end

if tokens >= n then
    tokens = tokens - n
    redis.call("HMSET", KEYS[1], "tokens", tostring(tokens), "last_refill_ms", tostring(now_ms))
    redis.call("PEXPIRE", KEYS[1], 60000)
    return {1, 0}
else
    local missing  = n - tokens
    local retry_ms = math.ceil((missing * 1000.0) / refill_per_sec)
    redis.call("HMSET", KEYS[1], "tokens", tostring(tokens), "last_refill_ms", tostring(now_ms))
    redis.call("PEXPIRE", KEYS[1], 60000)
    return {0, retry_ms}
end
